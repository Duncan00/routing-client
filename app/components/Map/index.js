import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import Error from '../../containers/HomePage/Error';

const { compose, withProps } = require('recompose');
const {
  withScriptjs,
  withGoogleMap,
  GoogleMap,
  DirectionsRenderer,
} = require('react-google-maps');

const Map = compose(
  withProps({
    googleMapURL: `https://maps.googleapis.com/maps/api/js?key=${
      process.env.GOOGLE_MAP_API_KEY
    }`,
    loadingElement: <div style={{ height: '100%' }} />,
    containerElement: <div style={{ width: '100%' }} />,
    mapElement: <div style={{ height: '100%' }} />,
  }),
  withScriptjs,
  withGoogleMap,
)(props => {
  // eslint-disable-next-line no-undef
  const { maps } = google;
  const DirectionsService = new maps.DirectionsService();
  const [directions, setDirections] = useState(null);
  const [error, setError] = useState(null);

  useEffect(() => {
    const { path } = props;

    setError(null);

    if (path) {
      const firstPoint = path[0];
      const origin = new maps.LatLng(firstPoint[0], firstPoint[1]);

      const lastPoint = path.slice(-1)[0];
      const destination = new maps.LatLng(lastPoint[0], lastPoint[1]);

      const waypoints = path
        .slice(1, -1)
        .map(([lat, lng]) => ({ location: new maps.LatLng(lat, lng) }));

      DirectionsService.route(
        {
          origin,
          destination,
          waypoints,
          travelMode: maps.TravelMode.DRIVING,
        },
        (result, status) => {
          if (status === maps.DirectionsStatus.OK) {
            setDirections(result);
          } else {
            setError(`Error fetching directions: ${result}`);
          }
        },
      );
    } else {
      setDirections(null);
    }
  }, [props.path]);

  return (
    <div>
      <GoogleMap
        defaultZoom={11}
        // Default center to Hong Kong
        defaultCenter={new maps.LatLng(22.28552, 114.15769)}
      >
        {directions && <DirectionsRenderer directions={directions} />}
      </GoogleMap>
      {error && <Error>{error}</Error>}
    </div>
  );
});

Map.propTypes = {
  path: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.string)),
};

export default Map;
