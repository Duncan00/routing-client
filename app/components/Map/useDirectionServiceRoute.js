import { useState, useEffect, useRef, useMemo, useCallback } from 'react';

/**
 * Whenever path updated, call google.maps.DirectionsService.route to get new directions
 * @param {Array} path
 * @returns {Array} [{data: {directions}, error}, updateDirections]
 */
export default function useDirectionServiceRoute(path) {
  // eslint-disable-next-line no-undef
  const { maps } = google;

  const DirectionsService = useMemo(() => new maps.DirectionsService(), []);

  const pathRef = useRef(path);

  const [directions, setDirections] = useState(null);
  const [error, setError] = useState(null);

  // Update directions whenever path changed
  const updateDirections = useCallback(async updatedPath => {
    setError(null);

    pathRef.current = updatedPath;

    if (updatedPath && updatedPath.length >= 2) {
      const firstPoint = updatedPath[0];
      const origin = new maps.LatLng(firstPoint[0], firstPoint[1]);

      const lastPoint = updatedPath.slice(-1)[0];
      const destination = new maps.LatLng(lastPoint[0], lastPoint[1]);

      const waypoints = updatedPath
        .slice(1, -1)
        .map(([lat, lng]) => ({ location: new maps.LatLng(lat, lng) }));

      DirectionsService.route(
        {
          origin,
          destination,
          waypoints,
          travelMode: maps.TravelMode.DRIVING,
        },
        (result, status) => {
          if (status === maps.DirectionsStatus.OK) {
            setDirections(result);
          } else {
            setError(`Error fetching directions: ${result}`);
          }
        },
      );
    } else {
      setDirections(null);
    }
  }, []);

  useEffect(() => {
    updateDirections(path);
  }, [path]);

  return [
    {
      data: { directions },
      error,
    },
    updateDirections,
  ];
}
