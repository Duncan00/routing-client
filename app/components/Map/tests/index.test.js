import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Map from '../index';

describe('<Map />', () => {
  it('should render Map container', () => {
    const path = [
      ['22.372081', '114.107877'],
      ['22.326442', '114.167811'],
      ['22.284419', '114.159510'],
    ];

    const renderer = new ShallowRenderer();
    const result = renderer.render(<Map path={path} />);

    expect(result).toMatchSnapshot();
  });
});
