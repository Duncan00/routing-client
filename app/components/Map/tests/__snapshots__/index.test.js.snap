// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Map /> should render Map container 1`] = `
<withGoogleMap(Component)
  containerElement={
    <div
      style={
        Object {
          "width": "100%",
        }
      }
    />
  }
  loadingElement={
    <div
      style={
        Object {
          "height": "100%",
        }
      }
    />
  }
  mapElement={
    <div
      style={
        Object {
          "height": "100%",
        }
      }
    />
  }
  path={
    Array [
      Array [
        "22.372081",
        "114.107877",
      ],
      Array [
        "22.326442",
        "114.167811",
      ],
      Array [
        "22.284419",
        "114.159510",
      ],
    ]
  }
/>
`;
